crates/ntx-builder/src/builder/data_store/mod.rs

```diff
@@ -0,0 +1,180 @@
+
+mod account_cache;
+
+// DATA STORE
+// =================================================================================================
+
+/// Network Transaction Builder data store.
+///
+/// [`NtxBuilderDataStore`] implements [`DataStore`] and provides all input data required by the NTB
+/// transaction executor. It maintains both a partial MMR and up-to-date account data. Accounts are
+/// kept fresh via an LRU cache, falling back to the backing store on a miss. The partial MMR is
+/// refreshed from the store whenever it becomes outdated.
+pub struct NtxBuilderDataStore {
+    mast_forest_store: TransactionMastStore,
+    store_client: StoreClient,
+    account_cache: NetworkAccountCache,
+    block_ref: Mutex<BlockHeader>,
+    partial_mmr: Mutex<PartialMmr>,
+}
+
+impl NtxBuilderDataStore {
+    /// Creates a new [`NtxBuilderDataStore`].
+    pub async fn new(store_client: StoreClient) -> Result<Self, NtxBuilderError> {
+        let account_cache = NetworkAccountCache::new(128.try_into().expect("not zero"));
+        let mast_forest_store = TransactionMastStore::new();
+
+        // SAFETY: OK to unwrap because passing `None` should return the latest data everytime
+        let (block_ref, partial_mmr) =
+            store_client.get_current_blockchain_data(None).await?.unwrap();
+
+        Ok(Self {
+            account_cache,
+            mast_forest_store,
+            store_client,
+            block_ref: Mutex::new(block_ref),
+            partial_mmr: Mutex::new(partial_mmr),
+        })
+    }
+
+    /// Inserts a note script's MAST forest into the MAST forest store.
+    pub fn insert_note_script_mast(&self, note_script: &NoteScript) {
+        self.mast_forest_store.insert(note_script.mast());
+    }
+
+    /// Retrieves the network account associated with a note tag and updates the state of the data
+    /// store accordingly to be able to execute transactions with the account.
+    /// Specifically, updating the state involves:
+    ///
+    /// - Retrieving the network account from the cache or through the Store and updating the cache
+    /// - Loading the account's MAST forest into the data store's MAST forest
+    pub async fn get_cached_acc_or_fetch_by_tag(
+        &self,
+        note_tag: NoteTag,
+    ) -> Result<Option<Account>, NtxBuilderError> {
+        let account_prefix: NetworkAccountPrefix = note_tag.try_into().unwrap();
+        // Look in cache, try the store otherwise
+        let account = if let Some(acc) = self.account_cache.get(account_prefix) {
+            Some(acc)
+        } else if let Some(acc) = self.store_client.get_network_account_by_tag(note_tag).await? {
+            // Insert to cache
+            self.account_cache.put(&acc)?;
+            Some(acc)
+        } else {
+            None
+        };
+
+        if let Some(acc) = &account {
+            self.mast_forest_store.insert(acc.code().mast());
+        }
+
+        Ok(account)
+    }
+
+    /// Updates the blockchain-related data by
```

Comment:
This doc looks unfinished.

Reply:
Good point, I think I copied it incorrectly. Changed!